pkg/network/tcp_peer.go

```diff
@@ -87,11 +87,18 @@ func (p *TCPPeer) putPacketIntoQueue(queue chan<- []byte, block bool, msg []byte
 	if !p.Handshaked() {
 		return errStateMismatch
 	}
+	var ret error
 	if block {
+		timer := time.NewTimer(p.server.TimePerBlock / 2)
```

Comment:
It's a nice optimization tested in docker environment, but does it affect the performance of the real network with 15-seconds blocks? I mean 7 seconds is a quite large timeout, are our peers *so* slow? Have you tried to connect to the mainnet and to log average duration of block message sending?

Reply:
> expected "normal" coefficients or 2/4/8 drastically degrade performance

Is it related to adaptive timeout?

Reply:
> It's a nice optimization

Not an optimization, I'd rather say it's a security fix.

> Have you tried to connect to the mainnet and to log average duration of block message sending

There is not a lot of things happening on mainnet, you just don't have any queuing issues there. A mainnet node can work with this patch, that's for sure.

> Is it related to adaptive timeout?

Yes. Adaptive one doesn't really work.